### PR TITLE
TS-4176: s3_auth add support for matrix params

### PR DESCRIPTION
**Pull request by robguima**
@shukitchan please review

https://issues.apache.org/jira/browse/TS-4176

### CODE REVIEW

File: plugins/experimental/s3_auth/s3_auth.cc [hunk 1]
```diff
@@ -346,26 +349,48 @@ S3Request::authorize(S3Config *s3)
...
+    con_md5 = TSMimeHdrFieldValueStringGet(_bufp, _hdr_loc, field_loc, -1, &con_md5_len);
+  }
+
+  // get the Content-Type if available - (buggy) clients may send it
+  // for GET requests too
+  field_loc = TSMimeHdrFieldFind(_bufp, _hdr_loc, TS_MIME_FIELD_CONTENT_TYPE, TS_MIME_LEN_CONTENT_TYPE);
+  if (field_loc) {
+    con_type = TSMimeHdrFieldValueStringGet(_bufp, _hdr_loc, field_loc, -1, &con_type_len);
+  }
+
```
Comment (shukitchan) [inline plugins/experimental/s3_auth/s3_auth.cc]: I think we should do a TSHandleMLocRelease on field_loc after each usage instead of one at the end.

**Comment by shukitchan** [comment]
@jpeach , can you take a look as well?

**Comment by shukitchan** [comment]
looks good now.

**Comment by robguima** [comment]
@shukitchan just to undo my formatting change from older patch - think someone did that between 5.3 and 6.x

**Comment by zwoop** [comment]
Please squash this down to one commit IMO.

**Comment by robguima** [comment]
@zwoop done thanks!

**Comment by zwoop** [comment]
One question: Can we be sure this doesn't break anything "backwards compatible" ? I.e. someone using the plugin today, it should just work the same with these changes?

**Comment by shukitchan** [comment]
I think we should use TSDebug as well. 

The change should be "backwards compatible".

**Comment by robguima** [comment]
Yes, we can be fairly sure it doesn't (break anything), because:
a) we've added functionality not changing the way it worked before (bug aside). So, if content-type, md5, and matrix params are not present, it should behave as before.
b) These fields not being present is actually the most common situation - so we've actually verified that the old functionality remains as is, because in most cases, md5, content-type, and matrix params are not used in GET requests.

The only caveat I can think of is using virtual host, which I am not and can't (very easily). Based on the code though, that little logic (prepending the host) seems unchanged by this PR.

This code is deployed in "production" right now in our aws region, and being tested routinely.

**Comment by zwoop** [comment]
Changing to TSDebug is fine, I don't recall why printf() was used, probably laziness :) I'll try to test this in our setup next week, but I agree that the changes ought to be backwards compatible.

**Comment by robguima** [comment]
@zwoop I will squash it to one commit if this change is OK.

**Comment by zwoop** [comment]
Couple of things on that last patch:

1) You have to run clang-format over the code, I'm fairly certain we would never indent like this:

 if (con_md5) TSDebug(PLUGIN_NAME, "%.*s", con_md5_len, con_md5);


2) Must we really introduce usage of std::string left("/"); ?? I guess this is inside a debug if() check, it's not a huge deal, but I cringe when I see std::string  :-).

**Comment by robguima** [comment]
2) hmm string is one of the few things I like about C++ :) especially for, like you said, code that is not in the main code path (e.g. debugging). It's just convenient and safer (null-termination) than using fixed size buffer with a bunch of checks - but your call.

**Comment by zwoop** [comment]
Yeah, the problem with it (and pretty much all of STL) is that it has hidden (and sometimes) unpredictable memory allocations behavior. I did some tests recently with bcall, where different compilers does different things when passing strings as argument to a function (one copied the string, the other did not).

I'm ok with std::string inside the if(debug tag) if you insist. My preference would be not to use std::string unless absolutely necessary (because it sets bad precedence), but it's difficult to make a strong argument here against it. :-).

**Comment by robguima** [comment]
@zwoop thought that was a pretty good argument myself. :) not a prob will change it - also forgot you guys are multiplat...

**Comment by robguima** [comment]
@zwoop @shukitchan please let me know if the std::string free version above is OK - if so I will squash it thx

**Comment by gtenev** [comment]
- Reviewed and tested the patch and it looks OK.
- Verified that it does not break backward compatibility, tested end-to-end against S3 with both virtual-hosted-style and path-style methods.
- Tested the Content-Type and Content-MD5 features and they worked correctly.
- Tested the matrix parameters feature, did not do end-to-end against S3 since we don't use matrix parameters in our setup but verified that the matrix parameters are added correctly to the CanonicalizedResource in the StringToSign and that the Authorization header is calculated correctly.

**Comment by shukitchan** [comment]
thanks... let me merge it tonight.

**Pull request closed without merging**